File: C#/OptimizationMethods/NumericUtils.cs
using System;

namespace OptimizationMethods
{
    public static class NumericUtils
    {
        private struct Mat2
        {
            public long m00;
            public long m01;
            public long m10;
            public long m11;
            public static Mat2 Identity() => new Mat2() { m00 = 1, m01 = 0, m10 = 0, m11 = 1 };
        }

        /// <summary>
        /// Конвертирует десятичную запись числа в рациональную, например, для числа 1.666 получим 1 2/3
        /// </summary>
        /// <param name="value">исходное число</param>
        /// <param name="rational_part">целые части</param>
        /// <param name="nominator">числитель</param>
        /// <param name="denominator">знаменатель</param>
        /// <param name="max_den">максимально допустимый знаменатель</param>
        public static void DecimalToRational(double value, out int rational_part, out int numerator, out int denominator, int max_den = 1000)
        {
            Mat2 m = Mat2.Identity();

            long ai;

            double x;

            int sign = value >= 0 ? 1 : -1;

            x = Math.Abs(value);
           
            long t;

            while (m.m10 * (ai = (long)x) + m.m11 <= max_den)
            {
                t = m.m00 * ai + m.m01;
            
                m.m01 = m.m00;
                m.m00 = t;
                
                t = m.m10 * ai + m.m11;
                
                m.m11 = m.m10;
                m.m10 = t;

                if (x == (double)ai) break; // AF: division by zero
                
                x = 1 / (x - (double)ai);
                
                if (x > (double)0x7FFFFFFF) break; // AF: representation failure
            }

            if ((rational_part = (int)(m.m00 / m.m10)) != 0)
            {
                numerator = (int)(m.m00 - rational_part * m.m10);

                rational_part *= sign;

                denominator = (int)m.m10;

                return;
            }
            rational_part = 0;

            numerator = (int)(sign * m.m00);

            denominator = (int)m.m10;
        }

        public static string ToRationalStr(double value, bool fullRational = true)
        {
            int r_part, num, denom;
            DecimalToRational(value, out r_part, out num, out denom);
            if (num    == 0)  return $"{r_part}";
            if (r_part == 0)  return $"{num}/{denom}";
            if (fullRational) return $"{(num + Math.Abs(r_part) * denom) * (r_part >= 0 ? 1 : -1)}/{denom}";
            return $"{r_part} {num}/{denom}";
        }

        public static string ToRationalStr(Vector vector, bool fullRational = true) => $"{{{string.Join(", ", vector.Map((v) => ToRationalStr(v, fullRational)))}}}";
    }
}

File: Java/OptimizationMethods/src/NumericUtils.java
public class NumericUtils
{
    /**
     * Конвертирует десятичную запись числа в рациональную,
     * например, 1.666 -> [1, 2, 3]
     *
     * Converts decimal representation of number into rational,
     * ex. 1.666 -> [1, 2, 3]
     *
     * @param value исходное число | original number
     * @param max_den максимально допустимый знаменатель | max denominator of number
     * @return int[]{integer_part, numerator, denominator}
     */
    public static int[] decimalToRational(double value, int max_den)
    {
         long m00 = 1;
         long m01 = 0;
         long m10 = 0;
         long m11 = 1;

        int[] number = new int[3];

        long ai;

        double x;

        int sign = value >= 0 ? 1 : -1;

        x = Math.abs(value);

        long t;

        while (m10 * (ai = (long)x) + m11 <= max_den)
        {
            t =m00 * ai + m01;

            m01 = m00;
            m00 = t;

            t = m10 * ai + m11;

            m11 = m10;
            m10 = t;
            if (x == (double)ai)
            {
                break;
            }   // AF: division by zero
            x = 1 / (x - (double)ai);
            if (x > (double)0x7FFFFFFF)
            {
                break;
            }  // AF: representation failure
        }

        if ((number[0] = (int)(m00 / m10)) != 0)
        {
            number[1] = (int)(m00 - number[0] * m10);

            number[0] *= sign;

            number[2] = (int)m10;

            return number;
        }
        number[0]  = 0;

        number[1]  = (int)(sign * m00);

        number[2]  = (int)m10;

        return number;
    }

    public static int[] decimalToRational(double value)
    {
        return decimalToRational(value,1000);
    }

    public static String toRationalStr(double value, boolean fullRational)
    {
        int[] number =  NumericUtils.decimalToRational(value);
        if (number[1] == 0)
        {
            return String.valueOf(number[0]);
        }
        if (number[0] == 0)
        {
            return String.valueOf(number[1]) + "/" + String.valueOf(number[2]);
        }

        if (fullRational)
        {
            return String.valueOf((number[1] + Math.abs(number[0]) * number[2]) * (number[0] >= 0 ? 1 : -1)) + "/" + String.valueOf(number[2]);
        }
        return String.valueOf(number[0]) + " " + String.valueOf(number[1]) + "/" + String.valueOf(number[2]);
    }

    public static String toRationalStr(double value)
    {
        return toRationalStr(value, true);
    }

    public static String toRationalStr(Vector value, boolean fullRational)
    {
        StringBuilder str = new StringBuilder("{ ");
        for (int i = 0; i < value.size() - 1; i++)
        {
            str.append(toRationalStr(value.get(i), fullRational));
            str.append(", ");
        }
        str.append(toRationalStr(value.get(value.size() - 1), fullRational));

        str.append(" }");
        return str.toString();
    }
    public static String toRationalStr(Vector value)
    {
        return toRationalStr( value,  true);
    }

}
